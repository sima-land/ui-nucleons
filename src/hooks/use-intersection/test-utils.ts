import { act } from '@testing-library/react';

/**
 * @deprecated Следует использовать утилиты из @sima-land/ui-nucleons/test-utils.
 */
export type IntersectionObserverEntryBase = Omit<
  IntersectionObserverEntry,
  'time' | 'rootBounds' | 'boundingClientRect' | 'intersectionRect'
>;

let intersectionMock: IntersectionMock | undefined;

/**
 * Возвращает новый объект, частично реализующий IntersectionObserverEntry.
 * @param base Базовая информация.
 * @return Объект.
 * @deprecated Следует использовать утилиты из @sima-land/ui-nucleons/test-utils.
 */
export function createFakeEntry(base: IntersectionObserverEntryBase): IntersectionObserverEntry {
  const entry: IntersectionObserverEntry = {
    ...base,

    time: Date.now(),

    /** @inheritdoc */
    get rootBounds() {
      // eslint-disable-next-line no-console
      console.error('Not implemented in mock');
      return null as any;
    },

    /** @inheritdoc */
    get boundingClientRect() {
      // eslint-disable-next-line no-console
      console.error('Not implemented in mock');
      return null as any;
    },

    /** @inheritdoc */
    get intersectionRect() {
      // eslint-disable-next-line no-console
      console.error('Not implemented in mock');
      return null as any;
    },
  } as unknown as IntersectionObserverEntry;

  return entry;
}

/**
 * Реализует интерфейс для управления состоянием пересечения элемента с область видимости.
 * @deprecated Следует использовать утилиты из @sima-land/ui-nucleons/test-utils.
 */
export class IntersectionMock {
  _original: typeof IntersectionObserver;
  _observers: FakeIntersectionObserver[] = [];

  /** Конструктор. */
  constructor() {
    this._original = window.IntersectionObserver;
  }

  /**
   * Подменяет глобальный конструктор IntersectionObserver.
   */
  apply() {
    if (intersectionMock) {
      throw Error('Already mocked');
    }

    // eslint-disable-next-line @typescript-eslint/no-this-alias
    intersectionMock = this;
    window.IntersectionObserver = FakeIntersectionObserver;
  }

  /**
   * Восстанавливает оригинальный глобальный конструктор IntersectionObserver.
   */
  restore() {
    if (window.IntersectionObserver !== FakeIntersectionObserver) {
      throw Error('Not mocked');
    }

    window.IntersectionObserver = this._original;
    intersectionMock = undefined;
  }

  /**
   * Имитирует изменение состояния пересечения элемента с областью видимости.
   * @param entryArg Данные пересечения элемента.
   */
  changeElementState(entryArg: IntersectionObserverEntryBase) {
    act(() => {
      const entry = createFakeEntry(entryArg);

      for (const observer of this._observers) {
        if (
          observer._elements.some(e => e === entry.target) &&
          observer.thresholds.includes(entry.intersectionRatio)
        ) {
          (observer as FakeIntersectionObserver)._runCallback(entry);
        }
      }
    });
  }

  /**
   * Записывает observer в список зарегистрированных.
   * @param observer Observer.
   */
  _registerObserver(observer: FakeIntersectionObserver) {
    this._observers.push(observer);
  }

  /**
   * Удаляет observer из списка зарегистрированных.
   * @param observer Observer.
   */
  _unregisterObserver(observer: FakeIntersectionObserver) {
    this._observers.splice(this._observers.indexOf(observer), 1);
  }
}

/**
 * Имитация IntersectionObserver.
 * @deprecated Следует использовать утилиты из @sima-land/ui-nucleons/test-utils.
 */
export class FakeIntersectionObserver implements IntersectionObserver {
  readonly root: Element | Document | null;
  readonly rootMargin: string;
  readonly thresholds: ReadonlyArray<number>;

  _callback: IntersectionObserverCallback;
  _elements: Element[];

  /** @inheritdoc */
  constructor(callback: IntersectionObserverCallback, options: IntersectionObserverInit = {}) {
    this.root = options?.root || null;
    this.rootMargin = options?.rootMargin || '0px 0px 0px 0px';
    this.thresholds = Array.isArray(options.threshold)
      ? options.threshold
      : [options.threshold || 0];

    this._callback = callback;
    this._elements = [];
    intersectionMock?._registerObserver(this);
  }

  /** @inheritdoc */
  observe(element: Element) {
    this._elements.push(element);
  }

  /** @inheritdoc */
  unobserve(element: Element) {
    this._elements = this._elements.filter(e => e !== element);
  }

  /** @inheritdoc */
  disconnect() {
    this._elements = [];
    intersectionMock?._unregisterObserver(this);
  }

  /** @inheritdoc */
  takeRecords() {
    // eslint-disable-next-line no-console
    console.error('Not implemented in mock');

    return [];
  }

  /**
   * Запускает callback.
   * @param entry Данные пересечения элемента.
   */
  _runCallback(entry: IntersectionObserverEntry) {
    this._callback([entry], this);
  }
}
