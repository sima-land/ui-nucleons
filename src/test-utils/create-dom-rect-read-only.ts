/**
 * Создает новый DOMRectReadOnly.
 * @inheritdoc
 */
export function createDOMRectReadOnly(
  x: number | undefined = 0,
  y: number | undefined = 0,
  width: number | undefined = 0,
  height: number | undefined = 0,
): DOMRectReadOnly {
  const rect = {
    x,
    y,
    width,
    height,
    top: y,
    left: x,
    bottom: y + height,
    right: x + width,
  };

  return {
    ...rect,

    // ВАЖНО: копируем чтобы не было возможности изменить результат
    toJSON: () => ({ ...rect }),
  };
}
