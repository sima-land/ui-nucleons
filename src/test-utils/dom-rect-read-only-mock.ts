/** */
export class DOMRectReadOnlyMock implements DOMRectReadOnly {
  /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/DOMRectReadOnly/bottom). */
  readonly bottom: number;

  /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/DOMRectReadOnly/height). */
  readonly height: number;

  /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/DOMRectReadOnly/left). */
  readonly left: number;

  /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/DOMRectReadOnly/right). */
  readonly right: number;

  /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/DOMRectReadOnly/top). */
  readonly top: number;

  /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/DOMRectReadOnly/width). */
  readonly width: number;

  /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/DOMRectReadOnly/x). */
  readonly x: number;

  /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/DOMRectReadOnly/y). */
  readonly y: number;

  /** @inheritdoc */
  constructor(
    x: number | undefined = 0,
    y: number | undefined = 0,
    width: number | undefined = 0,
    height: number | undefined = 0,
  ) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
    this.top = y;
    this.left = x;
    this.bottom = y + height;
    this.right = x + width;
    Object.freeze(this);
  }

  /** @inheritdoc */
  toJSON(): any {
    return {
      x: this.x,
      y: this.y,
      width: this.width,
      height: this.height,
      top: this.top,
      left: this.left,
      bottom: this.bottom,
      right: this.right,
    };
  }
}
