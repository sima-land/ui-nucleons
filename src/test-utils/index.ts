export { IntersectionObserverMock } from './intersection-observer-mock';
export { ResizeObserverMock } from './resize-observer-mock';
export { createDOMRectReadOnly } from './create-dom-rect-read-only';

// @todo matchMedia mock utils
