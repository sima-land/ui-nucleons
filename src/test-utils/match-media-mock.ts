const privates = new Map<MediaQueryListMock, { matches: boolean }>();

/**
 * Имитация window.matchMedia.
 */
export class MatchMediaMock {
  protected readonly registry: Set<MediaQueryListMock>;

  /**
   * @inheritdoc
   */
  constructor() {
    this.registry = new Set();
    this.matchMedia = this.matchMedia.bind(this);
  }

  /**
   * Имитация window.matchMedia.
   * @inheritdoc
   */
  matchMedia(query: string): MediaQueryListMock {
    const mql = new MediaQueryListMock(query, false);

    this.registry.add(mql);

    return mql;
  }

  /**
   * Меняет состояние созданных MediaQueryListEventMock и вызывает на них событие MediaQueryListEventMock с типом change.
   * @inheritdoc
   */
  simulateChange({ query, matches }: { query: string; matches: boolean }) {
    this.registry.forEach(mql => {
      if (mql.media === query) {
        privates.set(mql, { matches });
        mql.dispatchEvent(new MediaQueryListEventMock('change', { media: query, matches }));
      }
    });
  }
}

/**
 * Имитация MediaQueryList.
 */
class MediaQueryListMock extends EventTarget implements MediaQueryList {
  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/MediaQueryList/matches).
   * @inheritdoc
   */
  get matches(): boolean {
    return privates.get(this)?.matches ?? false;
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/MediaQueryList/media).
   */
  readonly media: string;

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/MediaQueryList/change_event).
   */
  onchange: ((this: MediaQueryList, ev: MediaQueryListEvent) => any) | null;

  /**
   * @inheritdoc
   */
  constructor(media: string, matches: boolean) {
    super();
    this.media = media;
    this.onchange = null;

    if (!privates.has(this)) {
      privates.set(this, { matches });
    }
  }

  /**
   * @inheritdoc
   */
  addEventListener<K extends keyof MediaQueryListEventMap>(
    type: K,
    listener: (this: MediaQueryList, ev: MediaQueryListEventMap[K]) => any,
    options?: boolean | AddEventListenerOptions,
  ): void;

  /**
   * @inheritdoc
   */
  addEventListener(
    type: string,
    listener: EventListenerOrEventListenerObject,
    options?: boolean | AddEventListenerOptions,
  ): void;

  /**
   * @inheritdoc
   */
  addEventListener(
    type: string,
    listener: EventListenerOrEventListenerObject,
    options?: boolean | AddEventListenerOptions,
  ) {
    super.addEventListener(type, listener, options);
  }

  /**
   * @inheritdoc
   */
  removeEventListener<K extends keyof MediaQueryListEventMap>(
    type: K,
    listener: (this: MediaQueryList, ev: MediaQueryListEventMap[K]) => any,
    options?: boolean | EventListenerOptions,
  ): void;

  /**
   * @inheritdoc
   */
  removeEventListener(
    type: string,
    listener: EventListenerOrEventListenerObject,
    options?: boolean | EventListenerOptions,
  ): void;

  /**
   * @inheritdoc
   */
  removeEventListener(
    type: string,
    listener: EventListenerOrEventListenerObject,
    options?: boolean | AddEventListenerOptions,
  ) {
    super.removeEventListener(type, listener, options);
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/MediaQueryList/addListener).
   * @deprecated
   * @inheritdoc
   */
  addListener(callback: ((this: MediaQueryList, ev: MediaQueryListEvent) => any) | null): void {
    if (callback) {
      this.addEventListener('change', callback);
    }
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/MediaQueryList/removeListener).
   * @deprecated
   * @inheritdoc
   */
  removeListener(callback: ((this: MediaQueryList, ev: MediaQueryListEvent) => any) | null): void {
    if (callback) {
      this.removeEventListener('change', callback);
    }
  }
}

/**
 * Имитация MediaQueryListEvent.
 */
class MediaQueryListEventMock extends Event implements MediaQueryListEvent {
  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/MediaQueryListEvent/matches).
   */
  readonly matches: boolean;

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/MediaQueryListEvent/media).
   */
  readonly media: string;

  /**
   * @inheritdoc
   */
  constructor(type: string, eventInitDict?: MediaQueryListEventInit) {
    super(type, eventInitDict);

    this.media = eventInitDict?.media ?? '';
    this.matches = eventInitDict?.matches ?? false;
  }
}
