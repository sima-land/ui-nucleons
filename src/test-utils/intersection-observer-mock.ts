import { createDOMRectReadOnly } from './create-dom-rect-read-only';

const privates = new Map<IntersectionObserverMock, { targets: Set<Element> }>();

/**
 * Имитация IntersectionObserver.
 */
export class IntersectionObserverMock implements IntersectionObserver {
  /**
   * Создает новый реестр экземпляров IntersectionObserverMock.
   * @return Реестр.
   */
  static createRegistry() {
    return new IntersectionObserverRegistry();
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/IntersectionObserver/root).
   */
  readonly root: Element | Document | null;

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/IntersectionObserver/rootMargin).
   */
  readonly rootMargin: string;

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/IntersectionObserver/thresholds).
   */
  readonly thresholds: ReadonlyArray<number>;

  protected readonly callback: IntersectionObserverCallback;
  protected readonly options: IntersectionObserverInit | undefined;

  /**
   * @inheritdoc
   */
  constructor(callback: IntersectionObserverCallback, options?: IntersectionObserverInit) {
    // ВАЖНО: имитируем поведение согласно реализации в браузере
    if (typeof callback !== 'function') {
      throw new TypeError('IntersectionObserverMock constructor: Argument 1 is not callable.');
    }

    // root
    this.root = options?.root ?? null;

    // rootMargin
    this.rootMargin = options?.rootMargin ?? '0px';

    // thresholds
    if (typeof options?.threshold === 'number') {
      this.thresholds = [options.threshold];
    } else if (Array.isArray(options?.threshold)) {
      this.thresholds = options.threshold;
    } else {
      this.thresholds = [0];
    }

    // specific for mock
    this.callback = callback;
    this.options = options;
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/IntersectionObserver/disconnect).
   * @inheritdoc
   */
  disconnect(): void {
    if (!privates.has(this)) {
      privates.set(this, { targets: new Set() });
    }

    privates.get(this)?.targets.clear();
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/IntersectionObserver/observe).
   * @inheritdoc
   */
  observe(target: Element): void {
    if (!privates.has(this)) {
      privates.set(this, { targets: new Set() });
    }

    privates.get(this)?.targets.add(target);
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/IntersectionObserver/takeRecords).
   * @inheritdoc
   */
  takeRecords(): IntersectionObserverEntry[] {
    // not implemented
    return [];
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/IntersectionObserver/unobserve).
   * @inheritdoc
   */
  unobserve(target: Element): void {
    if (!privates.has(this)) {
      privates.set(this, { targets: new Set() });
    }

    privates.get(this)?.targets.delete(target);
  }

  /**
   * Имитирует изменение состояния наблюдаемого элемента.
   * @param entry Данные состояния.
   */
  simulateEntryChange({
    target,
    isIntersecting,
    intersectionRatio,
  }: Pick<IntersectionObserverEntry, 'intersectionRatio' | 'isIntersecting' | 'target'>) {
    if (!privates.has(this)) {
      privates.set(this, { targets: new Set() });
    }

    if (privates.get(this)?.targets.has(target) && this.thresholds.includes(intersectionRatio)) {
      this.callback(
        [
          {
            target,
            isIntersecting,
            intersectionRatio,

            // not implemented
            time: Date.now(),
            boundingClientRect: createDOMRectReadOnly(),
            intersectionRect: createDOMRectReadOnly(),
            rootBounds: createDOMRectReadOnly(),
          },
        ],
        this,
      );
    }
  }
}

/**
 * Реестр экземпляров IntersectionObserverMock.
 */
export class IntersectionObserverRegistry {
  protected readonly items: Set<IntersectionObserverMock>;

  /**
   * @inheritdoc
   */
  constructor() {
    this.items = new Set();

    this.getObserver = this.getObserver.bind(this);
    this.simulateEntryChange = this.simulateEntryChange.bind(this);
  }

  /**
   * Создает, регистрирует и возвращает новый экземпляр IntersectionObserverMock.
   * @param params Аргументы IntersectionObserverMock.
   * @return Экземпляр IntersectionObserverMock.
   */
  getObserver(...params: ConstructorParameters<typeof IntersectionObserverMock>) {
    const item = new IntersectionObserverMock(...params);

    this.items.add(item);

    return item;
  }

  /**
   * Имитирует изменение состояния наблюдаемого элемента на всех экземплярах в реестре.
   * @param entry Данные состояния.
   */
  simulateEntryChange(
    entry: Pick<IntersectionObserverEntry, 'intersectionRatio' | 'isIntersecting' | 'target'>,
  ) {
    this.items.forEach(item => item.simulateEntryChange(entry));
  }
}
