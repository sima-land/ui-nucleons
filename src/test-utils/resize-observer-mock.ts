import { DOMRectReadOnlyMock } from './dom-rect-read-only-mock';

const privates = new Map<ResizeObserverMock, { targets: Set<Element> }>();

/**
 * Имитация ResizeObserver.
 */
export class ResizeObserverMock implements ResizeObserver {
  /**
   * Создает новый реестр экземпляров ResizeObserverMock.
   * @return Реестр.
   */
  static createRegistry() {
    return new ResizeObserverRegistry();
  }

  protected readonly callback: ResizeObserverCallback;

  /**
   * @inheritdoc
   */
  constructor(callback: ResizeObserverCallback) {
    // ВАЖНО: имитируем поведение согласно реализации в браузере
    if (typeof callback !== 'function') {
      throw new TypeError('ResizeObserverMock constructor: Argument 1 is not callable.');
    }

    // specific for mock
    this.callback = callback;
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/ResizeObserver/disconnect).
   * @inheritdoc
   */
  disconnect(): void {
    if (!privates.has(this)) {
      privates.set(this, { targets: new Set() });
    }

    privates.get(this)?.targets.clear();
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/ResizeObserver/observe).
   * @inheritdoc
   */
  observe(target: Element): void {
    if (!privates.has(this)) {
      privates.set(this, { targets: new Set() });
    }

    privates.get(this)?.targets.add(target);
  }

  /**
   * [MDN Reference](https://developer.mozilla.org/docs/Web/API/ResizeObserver/unobserve).
   * @inheritdoc
   */
  unobserve(target: Element): void {
    if (!privates.has(this)) {
      privates.set(this, { targets: new Set() });
    }

    privates.get(this)?.targets.delete(target);
  }

  /**
   * Имитирует изменение состояния наблюдаемого элемента.
   * @param entry Данные состояния.
   */
  simulateEntryChange({ target }: Pick<ResizeObserverEntry, 'target'>) {
    if (!privates.has(this)) {
      privates.set(this, { targets: new Set() });
    }

    if (privates.get(this)?.targets.has(target)) {
      this.callback(
        [
          {
            target,

            // not implemented
            borderBoxSize: [],
            contentBoxSize: [],
            contentRect: new DOMRectReadOnlyMock(),
            devicePixelContentBoxSize: [],
          },
        ],
        this,
      );
    }
  }
}

/**
 * Реестр экземпляров ResizeObserverMock.
 */
export class ResizeObserverRegistry {
  protected readonly items: Set<ResizeObserverMock>;

  /**
   * @inheritdoc
   */
  constructor() {
    this.items = new Set();
    this.getObserver = this.getObserver.bind(this);
    this.simulateEntryChange = this.simulateEntryChange.bind(this);
  }

  /**
   * Создает, регистрирует и возвращает новый экземпляр ResizeObserverMock.
   * @param params Аргументы ResizeObserverMock.
   * @return Экземпляр ResizeObserverMock.
   */
  getObserver(...params: ConstructorParameters<typeof ResizeObserverMock>) {
    const item = new ResizeObserverMock(...params);

    this.items.add(item);

    return item;
  }

  /**
   * Имитирует изменение состояния наблюдаемого элемента на всех экземплярах в реестре.
   * @param entry Данные состояния.
   */
  simulateEntryChange(entry: Pick<ResizeObserverEntry, 'target'>) {
    this.items.forEach(item => item.simulateEntryChange(entry));
  }
}
