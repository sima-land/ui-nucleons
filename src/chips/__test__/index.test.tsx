import { act, Simulate } from 'react-dom/test-utils';
import { render } from '@testing-library/react';
import { Chips, ChipsItemProps } from '..';

describe('Chips', () => {
  it('should renders correctly', async () => {
    const spy = jest.fn();

    const items: ChipsItemProps[] = [
      { children: 'Ручка для пакета' },
      { children: 'Спортивные кружки' },
      { children: 'Пакеты пищевые' },
      { children: 'Пакеты для заморозки' },
      { children: 'Домашние животные', checked: true },
      { children: 'Дикие звери' },
      { children: 'Крафт пакеты' },
      { children: 'Интерьер' },
      { children: 'Зоотовары' },
    ];

    const { container, findAllByTestId } = render(
      <Chips>
        {items.map((item, index) => (
          <Chips.Item key={index} {...item} onClick={spy} />
        ))}
      </Chips>,
    );

    expect(container).toMatchSnapshot();
    expect(await findAllByTestId('chips')).toHaveLength(1);

    expect(spy).toHaveBeenCalledTimes(0);

    await act(async () => {
      Simulate.click((await findAllByTestId('chips:item'))[4]);
    });

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should handle items with "href" property', async () => {
    const items: ChipsItemProps[] = [
      { children: 'Ручка для пакета' },
      { children: 'Спортивные кружки' },
      { children: 'Пакеты пищевые' },
    ].map(item => ({ ...item, href: 'www.foo-bar.com' }));

    const { container } = render(
      <Chips>
        {items.map((item, index) => (
          <Chips.Item key={index} {...item} />
        ))}
      </Chips>,
    );

    expect(
      container.querySelector<HTMLAnchorElement>('a[data-testid="chips:item"]')?.href,
    ).toContain('foo-bar');
  });

  it('should render with cross', () => {
    const items: ChipsItemProps[] = [
      { children: 'Ручка для пакета' },
      { children: 'Спортивные кружки' },
      { children: 'Пакеты пищевые' },
    ].map(item => ({ ...item, withCross: true }));

    const { container } = render(
      <Chips>
        {items.map((item, index) => (
          <Chips.Item key={index} {...item} />
        ))}
      </Chips>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should provide able to pass target props to anchor', async () => {
    const items: ChipsItemProps[] = [
      { href: 'https://www.foo.com', children: 'Foo', target: '_blank' },
      { href: 'https://www.bar.com', children: 'Bar', target: '_parent' },
    ];

    const { container } = render(
      <Chips>
        {items.map((item, index) => (
          <Chips.Item key={index} {...item} />
        ))}
      </Chips>,
    );

    const chips = container.querySelectorAll<HTMLAnchorElement>('a[data-testid="chips:item"]');

    expect(chips).toHaveLength(2);

    chips.forEach((elem, index) => {
      expect(elem.target).toBe(items[index].target);
    });
  });

  it('should handle "data-testid"', () => {
    const { queryAllByTestId } = render(
      <Chips data-testid='custom-chips'>
        <Chips.Item>Hello</Chips.Item>
        <Chips.Item>World</Chips.Item>
      </Chips>,
    );

    expect(queryAllByTestId('chips')).toHaveLength(0);
    expect(queryAllByTestId('custom-chips')).toHaveLength(1);
  });
});
