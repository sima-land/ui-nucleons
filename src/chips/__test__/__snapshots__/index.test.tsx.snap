// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips should render with cross 1`] = `
<div>
  <div
    class="list"
    data-testid="chips"
  >
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Ручка для пакета
      </span>
      <svg
        class="cross-svg"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.707 4.707a1 1 0 0 0-1.414-1.414L8 6.586 4.707 3.293a1 1 0 0 0-1.414 1.414L6.586 8l-3.293 3.293a1 1 0 1 0 1.414 1.414L8 9.414l3.293 3.293a1 1 0 0 0 1.414-1.414L9.414 8l3.293-3.293Z"
        />
      </svg>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Спортивные кружки
      </span>
      <svg
        class="cross-svg"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.707 4.707a1 1 0 0 0-1.414-1.414L8 6.586 4.707 3.293a1 1 0 0 0-1.414 1.414L6.586 8l-3.293 3.293a1 1 0 1 0 1.414 1.414L8 9.414l3.293 3.293a1 1 0 0 0 1.414-1.414L9.414 8l3.293-3.293Z"
        />
      </svg>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Пакеты пищевые
      </span>
      <svg
        class="cross-svg"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.707 4.707a1 1 0 0 0-1.414-1.414L8 6.586 4.707 3.293a1 1 0 0 0-1.414 1.414L6.586 8l-3.293 3.293a1 1 0 1 0 1.414 1.414L8 9.414l3.293 3.293a1 1 0 0 0 1.414-1.414L9.414 8l3.293-3.293Z"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`Chips should renders correctly 1`] = `
<div>
  <div
    class="list"
    data-testid="chips"
  >
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Ручка для пакета
      </span>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Спортивные кружки
      </span>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Пакеты пищевые
      </span>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Пакеты для заморозки
      </span>
    </a>
    <a
      class="item checked"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Домашние животные
      </span>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Дикие звери
      </span>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Крафт пакеты
      </span>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Интерьер
      </span>
    </a>
    <a
      class="item"
      data-testid="chips:item"
      role="button"
    >
      <span
        class="text"
      >
        Зоотовары
      </span>
    </a>
  </div>
</div>
`;
